### ISSUE TITLE
Unknown memory leak w/ don't keep activities

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/firefox-tv#2291]
## Steps to reproduce
- Enable don't keep activities (emulator)
- Open app
- Go to home screen
- Open app
- Open profiler in AS
- Force GC
- Dump heap
- In results, search for `MainActivity`

### Expected behavior
1 instance is retained

### Actual behavior
2 instances are retained. I have also confirmed that LeakCanary detects these leaks but unfortunately it doesn't provide a path to the leaks: "X was never GCed but no leak found" (maybe I need to explicitly annotate the Fragments?).

### Device information
* Fire TV device: emulator
* Latest affected Firefox version: master 05e595355d26dc4047ec9a9dc614075dfcd6dfe2
* Earliest affected Firefox version: ?

---

We've landed memory leaks before #1628 and so have our dependencies #1761 so, as a defensive measure, we really need to be testing for memory leaks in an automated way: we should schedule #797 (and perhaps write an integration test for using it) soon.

_edit: If you want to know details about how we fixed our most recent memory leak, see #1628, especially https://github.com/mozilla-mobile/firefox-tv/issues/1628#issuecomment-498443356 which includes extra information._

**Comment by athomasmoz** [comment]
@mcomella would this scenario ever happen in the wild? Or is it just with emulator

**Comment by mcomella** [comment]
This can happen in the wild: when the Android system cleans up our app to
save memory and recreates it when users try to access it again, users may
[1] be using more memory than they need to, which can eventually lead to an
out-of-memory application crash.

It's hard to say how often this happens: the out-of-memory crashes can
happen in many places so they're not aggregated into the same crash report;
it's also hard to distinguish them from other out-of-memory crashes with
different causes.

Memory leaks are better to address sooner rather than later:
- We don't know how often users are affected. However, they're notorious
for being the top Android crashers (though it's worse on phone where
rotation causes a clean up and recreate)
- When affected, users crash randomly, possibly during critical workflows
- If there is a memory leak, it becomes impossible to prevent additional
leaks
- The more memory leaks there are, the harder it is to debug them
- A previous memory leak was in components and, if applicable, the longer
we take to report it to them, the harder it will be to fix

[1]: The system can clean up our app a few different ways: only some types
of clean up will have the problem.

**Comment by liuche** [comment]
Try the old version of LeakCanary, apply explicit Fragments.

**Comment by Baron-Severin** [comment]
I wasn't able to make any real headway.  I tested:
- Rolling A-C back through 49
- Changing `EngineViewCache` to release references in `onStop`

The only other useful note is that `WebRenderFragment` is also leaked, but all of its references chain back to `MainActivity`.  So that confirms that it's the activity that's our problem.

**Comment by mcomella** [comment]
I paired with Severin. LeakCanary could not tell us the source of the leak (once we watched Fragments, it told us a Toast was referencing MainActivity but I've seen that before and I don't think it was the source of the problem).

I tried bisecting by building via the CL (`install*debug`), opening the profiler, forcing GC, and heap dumping and I was surprised to find every commit I pulled up had a leak, even the ones where I supposedly fixed previous leaks. Here are my notes, which get progressively less coherent as I bisecting:
```
78feb6cf27d9ba9b551dd40f3e9cec958aa8d7de - BAD
113072f1 - BAD
e673e06f15721f5fbae83c4b76bf77c220462241 - BAD
99942c79a3da973ef407b0847711e53d01245af7 - BAD
4bde5b12a00fa51e7ba2fcc2828c709d1b1edc7f - BAD ???
23a1288990e22a3fc06a2975bac0c307775a9e8d - 

5374c434 FEB 1

f0914cc10862b02e1ae727daecf3831bba48cc45
ec097ab058862129f0b5baa0294983ee32a67ab5 - BAD AUG 2018
```

It might be worth redoing this given the suspicious circumstances of everything forcing a leak.

**Comment by Baron-Severin** [comment]
> LeakCanary could not tell us the source of the leak (once we watched Fragments, it told us a Toast was referencing MainActivity but I've seen that before and I don't think it was the source of the problem).

We did also investigate this the toast reference, and it looked like it was handling context properly.

**Comment by mcomella** [comment]
I resumed bisecting. **Disclaimer:** as mentioned above, the methodology I'm using produced suspicious results.

I confirmed that 0fda8e23bf6246a36f5eafd67f824233a65c7b33, from July 2018, did not contain a leak. This is before we supposedly found (Dec 2018) and fixed (Feb 2019) the memory leak in https://github.com/mozilla-mobile/firefox-tv/issues/1628

As for our suspicious results, the cause may be:
- My methodology produces incorrect results
- We didn't actually fix the memory leak by fb9bd31bfc6a5708d2a84aa71c40f114fc741b90
- The dependencies we're pulling have changed upstream (I don't think maven is strict that published dependencies cannot be unpublished), meaning we fixed the leak at the time but now we're fetching new dependencies with the leak
- There is a bug in the AS profiler that's producing misleading results

**Comment by Baron-Severin** [comment]
This is caused by an issue in A-C.  Opened https://github.com/mozilla-mobile/android-components/issues/3254, which includes a detailed description of the problem.

**Comment by mcomella** [comment]
This will land in a-c 0.56.

**Comment by pocmo** [comment]
0.56.0 (+ 0.56.1) was released today!

**Comment by mcomella** [comment]
As a reviewer, I confirmed the memory leak was fixed in PR https://github.com/mozilla-mobile/firefox-tv/pull/2422 which upgrades a-c for the relman checklist #2402.

**Comment by Baron-Severin** [comment]
Closed in #2422
